### ISSUE TITLE
go: compressStep should use num, ok := map[string(byteslice)] to allow fast byteslice->string zero allocation compiler logic

### ISSUE THREAD

**Issue by odeke-em** [cosmos/ics23#109]
From a supply chain audit of this repo for the cosmos-sdk, we can see this code in compressStep https://github.com/confio/ics23/blob/a4daeb4c24ce1be827829c0841446abc690c4f11/go/compress.go#L92-L99

and notice on line we make a string `sig := string(bz)` then perform the lookup on line.

## Advice
Instead of unconditionally incurring the byteslice->string conversion on map hits, please use map[string(byteslice)] and only make that conversion when adding elements to the map so

```diff
diff --git a/go/compress.go b/go/compress.go
index 1064230..53d894c 100644
--- a/go/compress.go
+++ b/go/compress.go
@@ -93,17 +93,16 @@ func compressStep(step *InnerOp, lookup *[]*InnerOp, registry map[string]int32)
        if err != nil {
                panic(err)
        }
-       sig := string(bz)
 
        // load from cache if there
-       if num, ok := registry[sig]; ok {
+       if num, ok := registry[string(bz)]; ok {
                return num
        }
 
        // create new step if not there
        num := int32(len(*lookup))
        *lookup = append(*lookup, step)
-       registry[sig] = num
+       registry[string(bz)] = num
        return num
 }
```

/cc @elias-orijtech

**Comment by tac0turtle** [comment]
I thoughts its cheaper to do the conversion once and reuse? I might be missing something

**Comment by elias-orijtech** [comment]
The Go compiler optimizes the pattern `map[string(<byte-slice>)]` to avoid the implied allocation, but only if the conversion in done in-line. See https://codereview.appspot.com/83740044 for the implementation.

**Comment by xerpa43** [comment]
Hi, I am new to this and would like to work on it.

**Comment by crodriguezvega** [comment]
> Hi, I am new to this and would like to work on it.

Thank you, @xerpa43. Yes, the issue is still available, so I will assign it now to you to signal that you will be working on it.

**Comment by crodriguezvega** [comment]
> @crodriguezvega Thank You, from what I understand I have to make the changes given as advice, don't I?

Yes, that should be it. Let us know if you have any questions while you work on it!

**Comment by xerpa43** [comment]
I have opened a PR, please check it.

**Comment by crodriguezvega** [comment]
Closed by #149
